ale_linters/cpp/clangtidy.vim

```diff
@@ -8,15 +9,51 @@ let g:ale_cpp_clangtidy_checks = get(g:, 'ale_cpp_clangtidy_checks', ['*'])
 " This will disable compile_commands.json detection.
 let g:ale_cpp_clangtidy_options = get(g:, 'ale_cpp_clangtidy_options', '')
 
+" Set this option to manually point to the build directory for clang-tidy.
+" This will disable all the other clangtidy_options, since compilation
+" flags are contained in the json
+let g:ale_cpp_clangtidy_builddir = get(g:, 'ale_cpp_clangtidy_builddir', '')
+
+let g:ale_cpp_clangtidy_builddirnames =
+            \ get(g:, 'ale_cpp_clangtidy_builddirnames', ['build', 'bin'])
+
 function! ale_linters#cpp#clangtidy#GetCommand(buffer) abort
     let l:check_list = ale#Var(a:buffer, 'cpp_clangtidy_checks')
     let l:check_option = !empty(l:check_list)
     \   ? '-checks=' . ale#Escape(join(l:check_list, ',')) . ' '
     \   : ''
     let l:user_options = ale#Var(a:buffer, 'cpp_clangtidy_options')
-    let l:extra_options = !empty(l:user_options)
-    \   ? ' -- ' . l:user_options
-    \   : ''
+    let l:user_builddir = ale#Var(a:buffer, 'cpp_clangtidy_builddir')
+
+    " Build directory has the priority if
+    " both builddir and builddirnames options are defined
+    if empty(l:user_builddir)
+        let l:builddir_names = ale#Var(a:buffer, 'cpp_clangtidy_builddirnames')
+        for l:name in l:builddir_names
+            let l:candidates = finddir(l:name, expand('#' . a:buffer.':p:h') . ';', -1)
```

Comment:
Have a look at the implementation of `ale#python#FindVirtualenv`. You can use the `ale#path#Upwards` function with the `isdirectory` function for implementing this. I suggest implementing this as a `ale#c#FindCompileCommands` function. It could be useful for other tools.